tests/test_core.py

```diff
@@ -7037,7 +7037,11 @@ def encode_utf8(data):
       self.assertPathsIdentical(os.path.abspath('src.cpp'), m['source'])
       seen_lines.add(m['originalLine'])
     # ensure that all the 'meaningful' lines in the original code get mapped
-    assert seen_lines.issuperset([6, 7, 11, 12])
+    # when optimizing, the binaryen optimizer may remove some of them (by inlining, etc.)
+    if is_optimizing(self.emcc_args):
+      assert seen_lines.issuperset([11, 12]), seen_lines
```

Comment:
For some reason the wasm backend still gets all these lines preserved.. but it runs wasm-opt too.. some I'm a little confused.

Reply:
Oh interesting, I saw it broke on fastcomp, fixed the test, and saw it worked on upstream too, but didn't notice that it wasn't broken on upstream too.

I'm not very familiar with how debug info works, but looks like it works very differently in the wasm backend path - it has some custom sections with debug info in them that come from LLVM, I guess? Binaryen doesn't parse them, so the debug info is passed through. That means it's probably not valid if we optimize - probably not valid even if we just read and write the file, if it's based on offsets?

Seems like that means we've always been stuck in `-O0`, at best, for the wasm backend path + debug info? cc @yurydelendik

Reply:
@kripken yes, if a tool does not know how to read DWARF debug section it shall remove it after any code section change. In the binaryen, we operate with source maps, so it's okay to kill these sections. If a tool does not know how to process/transform the associated source map, the map becomes invalid and cannot be used with the wasm, but, I think, binaryen can do it.

Reply:
Interesting, thanks. So emscripten will definitely pass the wasm from wasm-ld into wasm-emscripten-finalize, which alters code sections (even without optimizations). So if the source map was generated before, it would be invalid, I think. When/where is it generated?

Reply:
I see the wasm-sourcemap.py util is called only at https://github.com/emscripten-core/emscripten/blob/6197e790d82440533fbf95350855dfb04671ada0/emscripten.py#L2188

Reply:
Thanks, I see. So it looks like we write the source map, then run `wasm-emscripten-finalize`, telling it the input source map, and it emits a modified one that is still valid. But later on we do run wasm-opt, which would break the source map. I'll fix that.